h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,169 @@
+  }
+
+  @Override
+  public void delete(Value v) {
+    BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    getClient().get(blobId).delete();
+  }
+
+  @Override
+  public void cleanUp() {
+    throw H2O.unimpl();
+  }
+
+
+  private static final Map<String, StringCache> keyCache = new HashMap<>();
+  private static final StringCache bucketCache = new StringCache() {
+    @Override
+    List<String> update() {
+      List<String> c = new ArrayList<>();
+      for (Bucket b : getClient().list().iterateAll()) {
+        c.add(GcsBlob.toCanonical(b.getName()));
+      }
+      return c;
+    }
+  };
+
+  @Override
+  public List<String> calcTypeaheadMatches(String filter, int limit) {
+    String input = GcsBlob.removePrefix(filter);
+    final String[] bk = input.split("/", 2);
+    if (bk.length == 1) {
+      return bucketCache.fetch(bk[0], limit);
+    } else if (bk.length == 2) {
+      if (!keyCache.containsKey(bk[0])) {
+        StringCache sc = new StringCache() {
+          @Override
+          List<String> update() {
+            List<String> cache = new ArrayList<>();
+            for (Blob b : getClient().get(bk[0]).list().iterateAll()) {
+              cache.add(GcsBlob.of(b).getCanonical());
+            }
+            return cache;
+          }
+        };
+        keyCache.put(bk[0], sc);
+      }
+      return keyCache.get(bk[0]).fetch(bk[1], limit);
+    }
+    return new ArrayList<>(0);
+  }
+
+  @Override
+  public void importFiles(String path,
+                          String pattern,
+                          ArrayList<String> files,
+                          ArrayList<String> keys,
+                          ArrayList<String> fails,
+                          ArrayList<String> dels) {
+    String bk[] = GcsBlob.removePrefix(path).split("/", 2);
+    Bucket bucket = getClient().get(bk[0]);
+    for (Blob blob : bucket.list().iterateAll()) {
+      if (bk.length == 1 || blob.getName().startsWith(bk[1])) {
+        GcsBlob gcsBlob = GcsBlob.of(blob.getBlobId());
+        try {
+          Key k = GcsFileVec.make(gcsBlob.getCanonical(), blob.getSize());
+          keys.add(k.toString());
+          files.add(gcsBlob.getCanonical());
+        } catch (Throwable t) {
+          fails.add(gcsBlob.getCanonical());
+        }
+      }
+    }
+  }
+
+  @Override
+  public InputStream open(String path) {
+    GcsBlob gcsBlob = GcsBlob.of(path);
+    final Blob blob = getClient().get(gcsBlob.getBlobId());
+    return new InputStream() {
+      @Override
+      public int read() throws IOException {
+        // very naive version with reading byte by byte
+        int res = -1;
+        try (ReadChannel reader = blob.reader()) {
+          ByteBuffer bytes = ByteBuffer.allocate(1);
+          while (reader.read(bytes) > 0) {
+            bytes.flip();
+            res = bytes.getInt();
+            bytes.clear();
+          }
+        } catch (IOException e) {
+          throw H2O.fail(e.getMessage(), e);
```

Comment:
`H2O.fail` will shut down the whole process. H2O.fail should typically be only used during H2O's boot. Once it is started, it should be avoided.

Reply:
Ok, I am changing to `FSIOException`.